server/src/main/java/jake2/server/SV_ENTS.java

```diff
@@ -143,219 +137,153 @@ void SV_EmitPacketEntities(client_frame_t from, client_frame_t to) {
-        client_frame_t frame, oldframe;
-        int lastframe;
-
-        //Com.Printf ("%i . %i\n", new
-        // Vargs().add(client.lastframe).add(sv.framenum));
         // this is the frame we are creating
-        frame = client.frames[gameImports.sv.framenum & Defines.UPDATE_MASK];
-        if (client.lastframe <= 0) { // client is asking for a retransmit
-            oldframe = null;
-            lastframe = -1;
-        } else if (gameImports.sv.framenum - client.lastframe >= (Defines.UPDATE_BACKUP - 3)) {
+        client_frame_t currentFrame = client.frames[gameImports.sv.framenum & Defines.UPDATE_MASK];
+        client_frame_t lastReceivedFrame;
+        int lastReceivedFrameNum;
+        // client is asking for a retransmit
+        if (client.lastReceivedFrame <= 0) {
+            lastReceivedFrame = null;
+            lastReceivedFrameNum = -1;
+        } else if (gameImports.sv.framenum - client.lastReceivedFrame >= (Defines.UPDATE_BACKUP - 3)) {
             // client hasn't gotten a good message through in a long time
             // Com_Printf ("%s: Delta request from out-of-date packet.\n",
             // client.name);
-            oldframe = null;
-            lastframe = -1;
-        } else { // we have a valid message to delta from
-            oldframe = client.frames[client.lastframe & Defines.UPDATE_MASK];
-            lastframe = client.lastframe;
+            lastReceivedFrame = null;
+            lastReceivedFrameNum = -1;
+        } else {
+            // we have a valid message to delta from
+            lastReceivedFrame = client.frames[client.lastReceivedFrame & Defines.UPDATE_MASK];
+            lastReceivedFrameNum = client.lastReceivedFrame;
         }
 
-        MSG.WriteByte(gameImports.msg, NetworkCommands.svc_frame);
-        MSG.WriteLong(gameImports.msg, gameImports.sv.framenum);
-        MSG.WriteLong(gameImports.msg, lastframe); // what we are delta'ing from
-        MSG.WriteByte(gameImports.msg, client.surpressCount); // rate dropped packets
-        client.surpressCount = 0;
+        new FrameMessage(
+                gameImports.sv.framenum,
+                lastReceivedFrameNum,
+                client.surpressCount,
+                currentFrame.areabytes,
+                currentFrame.areabits
+        ).writeTo(gameImports.msg);
 
-        // send over the areabits
-        MSG.WriteByte(gameImports.msg, frame.areabytes);
-        SZ.Write(gameImports.msg, frame.areabits, frame.areabytes);
+        client.surpressCount = 0;
 
         // delta encode the playerstate
-        SV_WritePlayerstateToClient(oldframe, frame, gameImports);
+        PlayerInfoMessage playerInfoMsg = buildPlayerInfoMessage(lastReceivedFrame != null ? lastReceivedFrame.ps : null, currentFrame.ps, gameImports);
+        playerInfoMsg.writeTo(gameImports.msg);
 
         // delta encode the entities
-        SV_EmitPacketEntities(oldframe, frame);
+        SV_EmitPacketEntities(lastReceivedFrame, currentFrame);
```

Comment:
todo: should also look like `PacketEntitiesMessage(...).writeTo(gameImports.msg)`